File: 백준/Silver/9465. 스티커/스티커.cc
#include<iostream>
#include<algorithm>

using namespace std;

int arr[2][100001];
int dp[3][100001];

int main() {

	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	int T;
	int n;
	int sol;
	int tmp;

	cin >> T;

	for (int t = 0; t < T; t++) {
		cin >> n;

		for (int i = 0; i < 2; i++) {
			for (int j = 1; j <= n; j++) {
				cin >> arr[i][j];
			}
		}

		dp[1][1] = arr[0][1];
		dp[2][1] = arr[1][1];

		for (int i = 2; i <= n; i++) {
			dp[0][i] = max(dp[1][i - 1], dp[2][i - 1]);
			dp[1][i] = max(dp[0][i - 1] + arr[0][i], dp[2][i - 1] + arr[0][i]);
			dp[2][i] = max(dp[0][i - 1] + arr[1][i], dp[1][i - 1] + arr[1][i]);
		}

		sol = max({ dp[0][n], dp[1][n], dp[2][n] });

		cout << sol << '\n';
	}


	return 0;
}

File: 백준/Silver/9465. 스티커/스티커.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int t = Integer.parseInt(br.readLine());

        // 테스트 케이스 t번 실행
        for (int i = 0; i < t; i++) {
            int n = Integer.parseInt(br.readLine());
            String[] str1 = br.readLine().split(" ");
            String[] str2 = br.readLine().split(" ");
            int[][] arr = new int[3][n+2];
            int[][] dp = new int[3][n+2];
            int[] dp2 = new int[n+2];
            int sol = 0;

            for (int j = 1; j <= n; j++) {
                arr[0][j] = Integer.parseInt(str1[j-1]);
                arr[1][j] = Integer.parseInt(str2[j-1]);
            }

            dp[0][1] = arr[0][1];
            dp[1][1] = arr[1][1];
            sol = Math.max(dp[0][1],dp[1][1]);
//            bw.write("dp[0]["+ 1 + "] = " + dp[0][1] + "\n");
//            bw.write("dp[1]["+ 1 + "] = " + dp[1][1] + "\n");
//            bw.write("sol= " + sol + "\n\n");

            for (int j = 2; j <= n; j++) {
                dp[0][j] = Math.max(arr[0][j]+dp[1][j-1], arr[0][j]+dp[1][j-2]);
                dp[1][j] = Math.max(arr[1][j]+dp[0][j-1], arr[1][j]+dp[0][j-2]);
                dp2[j] = Math.max(dp[0][j], dp[1][j]);
                sol = Math.max(sol,dp2[j]);

//                bw.write("dp[0]["+ j + "] = " + dp[0][j] + "\n");
//                bw.write("dp[1]["+ j + "] = " + dp[1][j] + "\n");
//                bw.write("sol= " + sol + "\n\n");
            }
            bw.write(sol + "\n");
            bw.flush();
        }
        bw.close();

    }
}